File: Problem21/main.hs
import Data.Numbers.Primes
import Data.List
import qualified Data.Map as M

main :: IO ()
main = print . sum . filter amicable $ [1 .. 9999]

amicable :: Integral a => a -> Bool
amicable n = n == d (d n) && n /= d n where
    d = sum . properDivisors

properDivisors :: Integral a => a -> [a]
properDivisors n = filter (/= n) $ divisors n

divisors :: Integral a => a -> [a]
divisors 0 = []
divisors n = map product . mapM (\(p, v) -> map (p ^) [0 .. v]) $ primeFactorization n

primeFactorization :: Integral a => a -> [(a, Int)]
primeFactorization n = zip (nub l) (map length $ group l) where
    l = if n == 0 then [] else primeFactors n

File: Problem21/main.java
import java.util.LinkedList;
import java.util.List;

public class main {
    public static void main(String[] args) {
        int sum = 0;
        for (int i = 1; i < 10000; i++) {
            if (isAmicable(i)) {
                sum += i;
            }
        }
        System.out.println(sum);
    }
    
    private static boolean isAmicable(int a) {
        int b = sum(properDivisors(a));
        if (sum(properDivisors(b)) == a && a != b) {
            return true;
        } else {
            return false;
        }
    }

    private static List<Integer> properDivisors(int n) {
        List<Integer> divisors = new LinkedList<Integer>();
        for (int i = 1; i <= n/2; i++) {
            if (n%i == 0) {
                divisors.add(i);
            }
        }
        return divisors;
    }

    private static int sum(List<Integer> list) {
        int sum = 0;
        for (Integer i : list) {
            sum += i;
        }
        return sum;
    }
}
